fix(runtime): let match(it+paren) skips use treesitter

emiasims opened the pull request:
Problem: When treesitter is enabled, by default syntax groups are not defined, but these
groups are used to identify where to skip matches in matchit and matchparen.

Solution: This patch does three things:
1. If syntax is enabled regardless of treesitter (`vim.bo.syntax='on'`):
   Use original implementation.
2. If treesitter is enabled and syntax is not:
   Match the syntax groups (i.e. `comment\|string`) against treesitter captures
   to check for skipped groups.
3. Add an explicit treesitter syntax for marking captures to skip:
   matchit uses `b:match_skip` to determine what counts as skippable
   Where 's:comment\|string' uses a match of the named syntax groups against
   a regex match of comment\|string, 't:comment\|string' now uses vim regex
   to match against the names of the treesitter capture groups.

ref #15462

justinmk reviewed (approved):
No objection, though https://github.com/neovim/neovim/issues/15462 can remain open to track a full Lua rewrite of matchit.

clason commented:
Can you explain (also in the commit message) what this patch does and how it improves the situation with treesitter (without regressing on legacy syntax)?

clason commented:
(And, yes, this would be a stopgap until the planned total rewrite, which cannot come too soon for me.)

emiasims committed:
fix(runtime): let matchit and matchparen skips fallback on treesitter captures

When treesitter is enabled, by default syntax groups are not defined, but these
groups are used to identify where to skip matches in matchit and matchparen.

This patch does three things:
1. If syntax is enabled regardless of treesitter (`vim.bo.syntax='on'`):
   Use original implementation.
2. If treesitter is enabled and syntax is not:
   Match the syntax groups (i.e. `comment\|string`) against treesitter captures
   to check for skipped groups.
3. Add an explicit treesitter syntax for marking captures to skip:
   matchit uses `b:match_skip` to determine what counts as skippable
   Where 's:comment\|string' uses a match of the named syntax groups against
   a regex match of comment\|string, 't:comment\|string' now uses vim regex
   to match against the names of the treesitter capture groups.

milisims commented:
When treesitter is enabled, by default syntax groups are not defined, but these
groups are used to identify where to skip matches in matchit and matchparen.

This patch does three things:
1. If syntax is enabled regardless of treesitter (`vim.bo.syntax='on'`): Use original implementation.
2. If treesitter is enabled and syntax is not: Match the syntax groups (i.e. `comment\|string`) against treesitter captures to check for skipped groups.
3. Add an explicit treesitter syntax for marking captures to skip: matchit uses `b:match_skip` to determine what counts as skippable Where 's:comment\|string' uses a match of the named syntax groups against a regex match of comment\|string, 't:comment\|string' now uses vim regex to match against the names of the treesitter capture groups.

clason commented:
Hmm, I'm not sure this is going to work in its current form: the expected captures don't match well to the standard captures we use: https://neovim.io/doc/user/treesitter.html#treesitter-highlight-groups

I'd feel much better if this PR came with tests, but the bundled plugins are unfortunately not covered.

milisims commented:
Comment and strings work very well, which is a huge improvement, and is the default for ftplugins that don’t define `b:match_skip`. Other languages use ‘special’ and ‘character’ like c, which works fine.

I went through every bundled ftplugin for `match_skip`, and I found:
- basic: custom match_skip, requires syntax. Probably could be edited to work pretty well with a test if ts is loaded and just defining a `t:` match skip.
- clojure: regex syn group is highlighted as string.regex which already works (the match isn’t anchored). string/comment work
- erlang: erlangmodifier and erlangquotedatom would possibly be ignored, I don’t know how they’re parsed or highlighted in the ts grammar. otherwise comment/string work
- julia: needs a `t:` definition
- perl: comment and string work, but others wouldn’t without a `t:`
- php: custom, uses syntax but could fallback at least to comment and strings. possibly just a `t:` def
- ruby: custom, but most could be covered with `t:` I think
- rust: would need a `t:` but easy, comment & string still work
- vim: needs a `t:`
Most of these could be improved from by simply falling back to comment/strings if ts is enabled and syntax is not.

These langs define match_skip but have no ts grammar: eiffel, mediawiki, mf, modula2, zimbu, zsh
At least, not in nvim-treesitter.

clason commented:
Yes, I am specifically concerned about the "subcaptures". For example, `@string.special` or `@comment.documentation` -- shouldn't these be covered, too?

milisims commented:
As is, the spec is just matched against the group. So `‘comment\|string` matches against `comment.documentation`, which is how the matchit implementation works, so subgroups are all matched. Should we anchor it?

clason commented:
No, then that should be fine.

justinmk commented:
https://github.com/neovim/neovim/pull/31861 is underway, should we merge this meanwhile? If not then let's close it.

clason commented:
Low risk, so let's just merge (and backport).
